### ISSUE TITLE
Deprecate InjectMocks

### ISSUE THREAD

**Issue by TimvdLippe** [mockito/mockito#1518]
I think no one of the core team is a fan of this feature. The implementation is not as full-fletched as other dependency injection mechanisms. At the time of writing this issue, there are 17 open issues describing various problems with `@InjectMocks`. Moreover, as it solves a problem unrelated to mocking and there are different tools available to solve this use case, I propose we deprecate `@InjectMocks` and schedule for removal in Mockito 3/4. (I think 4 is best, to keep 3 a minimal breaking change version to ease upgrades).

**Comment by ChristianSchwarz** [comment]
Good idea! 
A lib relate to this topic: [Jukito ](https://github.com/ArcBees/Jukito) handles the dependency injection for Mockito using Guice.

**Comment by bric3** [comment]
I agree however, we should have a proper replacement already in place.
By the way dropping this may bring havoc in the world.

**Comment by TimvdLippe** [comment]
For simple cases, manual constructing the object should work. For more complicated cases, I think Jukito is appropriate.

An option is to, in the deprecation message, ask for feedback if either of these solutions is not feasible for the user. That way we know if there are actually use cases we can't cover (and thus would have to keep supporting `@InjectMocks`)

**Comment by mockitoguy** [comment]
Our users love this feature! (just talk to them, you'll see :)

We already had this discussion before the feature was introduced. @bric3, do you remember? It was contentious back then, we knew that we cannot compete with fully fledged dependency injection framework. We documented this and moved on. Over time, it paid off because I heard multiple times that our users really like this feature.

Does deprecating this feature helps our community? There are many users (like myself ;) who are happy with the current state of InjectMocks. I have never had problems with it. I have not seen a request from the community to get rid of this feature.

@TimvdLippe, what is the problem you're are trying to solve? It's ok to have 17 tickets - users can help us fix them. We can also close those tickets if we don't want the fixes.

**Comment by TimvdLippe** [comment]
The problem with all these issues is that the expectation of the user is broken. By having an inheritently broken feature in our API, we are not living up to our promises imo. Rather than having something that sometimes work, I think it is better to let the task be solved by those frameworks that are built to solve them.

**Comment by msangel** [comment]
[Jukito](https://github.com/ArcBees/Jukito) seems to be a good choice for google-guice based apps. How about spring boot autoconfiguration? Will it work?

**Comment by bric3** [comment]
@mockitoguy Yes I do remember, you were quite lucid back then and I was quite young.
@TimvdLippe I do not think it is _inherently broken_ per se, but _inherently limited_. The same I don't think it _sometimes work_, but rather that it _sometimes doesn't work_.

I think the two system can live together, and complement each other. Most problems arise for more complicated object graph which is clearly not advertised as what Mockito injection is meant to do.

I'm not against moving it to another jar. Nor to deprecate it, but a replacement should be available.
My opinion is that I liked about Mockito is the completeness of the jar (and it's dependencies) ; I don't have to declare yet another dependency.

**Comment by mockitoguy** [comment]
>By having an inherently broken feature in our API, we are not living up to our promises imo.

It is a useful feature that has limitations we are not interested in addressing. Perhaps we should work on better error reporting / early failure when InjectMocks hits one of the limitations?

Do we aspire to build a "perfect" tool? We want to build a useful tool, that is "great enough". In many other parts of our API we have caveats, typically caused by Java limitations. We always optimized to make the 95% case experience great, and remaining 5% decent (or not horrible ;) - via good error messages, great documentation. Given Mockito popularity, this strategy has worked out.

Does removing this feature make Mockito better? I have a talk about Mockito 3 at CodeOne next week. I'll ask the audience about InjectMocks.

Nice discussion! Thank you everybody for chiming in!

**Comment by glureau-betclic** [comment]
Did you receive any feedback @mockitoguy ?
